README.md
# Swift PostgreSQL adapter

PostgreSQL adapter for MRI Ruby.

NOTE: This has nothing to do with Swift programming language (OSX, iOS)

## Features

* Lightweight & fast
* Result typecasting
* Prepared statements
* Asynchronous support using PQsendQuery family of functions
* Nested transactions

## Requirements

* postgresql client deveopment libraries (libpq-dev)
* uuid development libraries (uuid-dev)

## Building

```
git submodule update --init
rake
```

## API

```
  Swift::DB::Postgres
    .new(options)
    #execute(sql, *bind)
    #prepare(sql)
    #begin(savepoint = nil)
    #commit(savepoint = nil)
    #rollback(savepoint = nil)
    #transaction(savepoint = nil, &block)
    #native_bind_format(&block)
    #native_bind_format=(value)
    #ping
    #close
    #closed?
    #escape(text)
    #query(sql, *bind)
    #fileno
    #result
    #write(table = nil, fields = nil, io_or_string)
    #read(table = nil, fields = nil, io = nil, &block)
    #encoder=
    #decoder=
    #typemap

  Swift::DB::Postgres::Statement
    .new(Swift::DB::Postgres, sql)
    #execute(*bind)
    #release

  Swift::DB::Postgres::Result
    #selected_rows
    #affected_rows
    #fields
    #types
    #each
    #insert_id
    #clear
    #get(row, column)
```

## Connection options

```
╭────────────────────╥────────────┬─────────────╮
│ Name               ║  Default   │  Optional   │
╞════════════════════╬════════════╪═════════════╡
│ db                 ║  -         │  No         │
│ host               ║  -         │  Yes        │
│ port               ║  5432      │  Yes        │
│ user               ║  Etc.login │  Yes        │
│ password           ║  nil       │  Yes        │
│ encoding           ║  utf8      │  Yes        │
│ ssl[:sslmode]      ║  allow     │  Yes        │
│ ssl[:sslcert]      ║  nil       │  Yes        │
│ ssl[:sslkey]       ║  nil       │  Yes        │
│ ssl[:sslrootcert]  ║  nil       │  Yes        │
│ ssl[:sslcrl]       ║  nil       │  Yes        │
└────────────────────╨────────────┴─────────────┘
```

## Bind parameters and hstore operators

`Swift::DB::Postgres` uses `?` as a bind parameter and replaces them with the `$` equivalents. This causes issues when
you try to use the HStore `?` operator. You can permanently or temporarily disable the replacement strategy as below:

```ruby

db.native_bind_format = true
db.execute("select * from users where tags ? $1", 'mayor')
db.native_bind_format = false

db.native_bind_format do
  db.execute("select * from users where tags ? $1", 'mayor')
end
```

## Example

```ruby
require 'swift/db/postgres'

db = Swift::DB::Postgres.new(db: 'swift_test', encoding: 'utf8')

db.execute('drop table if exists users')
db.execute('create table users (id serial, name text, age integer, created_at timestamp)')
db.execute('insert into users(name, age, created_at) values(?, ?, ?)', 'test', 30, Time.now.utc)

row = db.execute('select * from users').first
p row #=> {:id => 1, :name => 'test', :age => 30, :created_at=> #<Swift::DateTime>}
```

### Typecast support

The library understands the following Ruby and Postgres types and will attempt to typecast
values in either direction appropriately. Types not listed below will be stringified on the
way in to the database and returned as a raw string value on the way out.

```
╭────────────────┬─────────────────────╮
│ Ruby           │  PostgreSQL         │
╞════════════════╪═════════════════════╡
│ TrueClass      │  boolean            │
│ FalseClass     │  boolean            │
│ Bignum         │  bigint             │
│ Fixnum         │  int                │
│ Integer        │  int                │
│ Float          │  float, double      │
│ BigDecimal     │  numeric            │
│ Time           │  timestamp          │
│ DateTime       │  timestamp          │
│ String         │  text               │
│ Symbol         │  text               │
│ IO             │  bytea              │
└────────────────┴─────────────────────┘
```

### Custom encoder

`Swift::DB::Postgres#execute` or `Swift::DB::Postgres::Statement#execute` will attempt to encode bind values and
then fallback to stringifying unsupported types using `Object#to_s`. This can be overriden with a customer encoder.

e.g.

```ruby
db = Swift::DB::Postgres.new(db: 'swift_test')
db.encoder = proc do |value|
  case value
    when Hash  then JSON.dump(value)
    when Array then ...
    else nil
  end
end
```

The encoder can be any object that implements `.call` and is passed the bind value. If the encoder returns
`nil` then `#execute` will fallback to stringifying the value.


### Custom decoder

Logical inverse of above. `Result#each` attempts to decode known types but will return a raw string if it cannot
decode a value. A decoder can be provided to extend the typecasting support, the decoder is passed in extra
metadata hints to help including a field name and the Postgres OID.

```ruby
db = Swift::DB::Postgres.new(db: 'swift_test')
db.decoder = proc do |field, oid, value|
  case oid
    when 114 then JSON.parse(value)
    when 142 then Nokogiri::XML(value)
    else nil
  end
end
```

You can get a list of statically generated OIDs with type description using `Swift::DB::Postgres#typemap` but
this method only exists for convenience and the type map OIDs may not include all the supported types in your
PostgreSQL instance.

### Asynchronous

There are several approaches to handling IO wait and concurrency but all of them require creating a connection
pool and checking out a connection to run queries in parallel.

1. Create a connection pool with separate connections to the database.
2. Use `Swift::DB::Postgres#query` to execute a SQL in a non-blocking fashion.
3. Use one of the polling schemes below to retrieve results.

#### Threads

Call `Swift::DB::Postgres::Result#each` in a thread to handle IO wait

```ruby
require 'swift/db/postgres'

rows = []
pool = 3.times.map {Swift::DB::Postgres.new(db: 'swift_test')}

3.times do |n|
  Thread.new do
    pool[n].query("select pg_sleep(#{(3 - n) / 10.0}), #{n + 1} as query_id") do |row|
      rows << row[:query_id]
    end
  end
end

Thread.list.reject {|thread| Thread.current == thread}.each(&:join)
p rows #=> [3, 2, 1]
```

#### IO select

Use `Swift::DB::Postgres#fileno` to get the unix file descriptor and create an IO object. You can
then use `Kernel#select` to poll the IO object.

```ruby
require 'swift/db/postgres'

rows = []
pool = 3.times.map {Swift::DB::Postgres.new(db: 'swift_test')}

io_list = pool.map(&:fileno).map {|fd| IO.for_fd(fd) }
io_lookup = Hash[io_list.map(&:fileno).zip(pool)]

3.times do |n|
  pool[n].query("select pg_sleep(#{(3 - n) / 10.0}), #{n + 1} as query_id")
end

loop do
  ready = IO.select(io_list, nil, nil, 0.5)
  break if ready.nil?
  ready.first.each do |io|
    io_lookup[io.fileno].result.each do |tuple|
      rows << tuple[:query_id]
    end
  end
end

p rows #=> [3, 2, 1]
```

#### EventMachine.watch

Similar to above example, you can use `EventMachine.watch` on the file descriptor returned by `Swift::DB::Postgres#fileno`

See https://www.rubydoc.info/github/eventmachine/eventmachine/EventMachine.watch

### Data I/O

The adapter supports data read and write via COPY command.

```ruby
require 'swift/db/postgres'

db = Swift::DB::Postgres.new(db: 'swift_test')
db.execute('drop table if exists users')
db.execute('create table users (id serial, name text)')

db.write('users', %w{name}, "foo\nbar\nbaz\n")
db.write('users', %w{name}, StringIO.new("foo\nbar\nbaz\n"))
db.write('users', %w{name}, File.open("users.dat"))


db.read('users', %w{name}) do |data|
  p data
end

csv = File.open('users.csv', 'w')
db.execute('copy users to stdout with csv')
db.read(csv)
```

## Performance

Tests:
* Insert 1000 rows
* Read them back 100 times with typecast enabled

Caveats:
* This is just a raw performance benchmark, each library has its advantages and disadvantages.

```
$ ruby -v
ruby 2.1.5p273 (2014-11-13 revision 48405) [x86_64-linux]

$ ruby check.rb
                           user     system      total        real
swift insert           0.020000   0.000000   0.020000 (  0.090353)
swift select           0.170000   0.000000   0.170000 (  0.268109)

do_postgres insert     0.020000   0.020000   0.040000 (  0.096619)
do_postgres select     0.550000   0.000000   0.550000 (  0.659646)

pg insert              0.020000   0.000000   0.020000 (  0.069376)
pg select              0.680000   0.030000   0.710000 (  0.819120)
```

## License

MIT


CHANGELOG
== 0.4.0 (2018-06-30)

* encoder and decoder support for types.
* Result#clear to free up memory instead of waiting for a GC cycle.
* default to using unix socket if host is not provided.

== 0.3.2 (2018-06-15)

* updated extconf.rb to use pg_config if present.

== 0.3.1 (2015-03-22)

* prepared statement uuid overflow fix. 

== 0.3.0 (2013-02-19)

* supports encoding option while connecting. 

== 0.2.7 (2012-11-22)

* rb_gc_register_address() the bind values.

== 0.2.6 (2012-11-14)

* Ruby 2.x build fixes.
* Pause GC around execute() with bind values.

== 0.2.5 (2012-09-26)

* allow more than 99 placeholders. contributor: @filterfish

== 0.2.4 (2012-08-18)

* gc protect bind args.

== 0.2.3 (2012-07-31)

* remove rb_protect from Result#each.

== 0.2.2 (2012-07-30)

* removed rb_gc_mark called outside mark phase.

== 0.2.1 (2012-07-30)

* rb_protect yielding result tuples.

== 0.2.0 (2012-07-29)

* Added Adapter#native_bind_format and Adapter#native_bind_format=

== 0.1.2 (2012-07-29)

* gc guard bind parameters during execute.

== 0.1.1 (2012-07-26)

* Updated compile flags.

== 0.1.0 (2012-07-20)

* Initial version.


ext/swift/db/postgres/typecast.c
// vim:ts=4:sts=4:sw=4:expandtab

// (c) Bharanee Rathna 2012

#include "common.h"
#include "typecast.h"
#include "datetime.h"

const static struct {
    int oid;
    const char *type;
} oid_types[] = {
    {16, "boolean"},
    {17, "bytea"},
    {18, "char"},
    {19, "name"},
    {20, "bigint"},
    {21, "smallint"},
    {23, "integer"},
    {25, "text"},
    {114, "json"},
    {142, "xml"},
    {600, "point"},
    {601, "lseg"},
    {602, "path"},
    {603, "box"},
    {604, "polygon"},
    {628, "line"},
    {650, "cidr"},
    {700, "real"},
    {701, "double precision"},
    {702, "abstime"},
    {703, "reltime"},
    {704, "tinterval"},
    {718, "circle"},
    {774, "macaddr8"},
    {790, "money"},
    {829, "macaddr"},
    {869, "inet"},
    {1007, "int array"},
    {1009, "text array"},
    {1042, "character"},
    {1043, "character varying"},
    {1082, "date"},
    {1083, "time without time zone"},
    {1114, "timestamp without time zone"},
    {1184, "timestamp with time zone"},
    {1186, "interval"},
    {1266, "time with time zone"},
    {1560, "bit"},
    {1562, "bit varying"},
    {1700, "numeric"},
    {2277, "anyarray"},
    {2283, "anyelement"},
    {2776, "anynonarray"},
    {2950, "uuid"},
    {3500, "anyenum"},
    {3614, "tsvector"},
    {3615, "tsquery"},
    {3642, "gtsvector"},
    {3802, "jsonb"},
    {3831, "anyrange"},
    {3904, "int4range"},
    {3906, "numrange"},
    {3908, "tsrange"},
    {3910, "tstzrange"},
    {3912, "daterange"},
    {3926, "int8range"}
};

#define date_parse(klass, data,len) rb_funcall(datetime_parse(klass, data, len), fto_date, 0)

ID fnew, fto_date, fstrftime;
VALUE cBigDecimal, cStringIO;
VALUE dtformat;
VALUE cDateTime;

#define TO_UTF8(value) rb_str_encode(value, rb_str_new2("UTF-8"), 0, Qnil)
#define UTF8_STRING(value) strcmp(rb_enc_get(value)->name, "UTF-8") ? TO_UTF8(value) : value

VALUE typecast_to_str(VALUE value) {
    return UTF8_STRING(rb_funcall(value, rb_intern("to_s"), 0));
}

VALUE typecast_encode(VALUE value) {
    switch (TYPE(value)) {
        case T_STRING:
            return UTF8_STRING(value);
        case T_TRUE:
            return rb_str_new2("1");
        case T_FALSE:
            return rb_str_new2("0");
        case T_FLOAT:
        case T_BIGNUM:
        case T_FIXNUM:
        case T_SYMBOL:
        case T_RATIONAL:
            return typecast_to_str(value);
        default:
            if (rb_obj_is_kind_of(value, rb_cTime) || rb_obj_is_kind_of(value, cDateTime))
                return rb_funcall(value, fstrftime, 1, dtformat);
            else if (rb_obj_is_kind_of(value, rb_cIO) || rb_obj_is_kind_of(value, cStringIO))
                return rb_funcall(value, rb_intern("read"), 0);
            else
                return Qnil;
    }
}

VALUE typecast_decode(const char *data, size_t size, int oid) {
    VALUE value;
    unsigned char *bytea;
    size_t bytea_len;
    switch (oid) {
        case 16:
            return (data && (data[0] =='t' || data[0] == '1')) ? Qtrue : Qfalse;
        case 17:
            bytea = PQunescapeBytea((const unsigned char*)data, &bytea_len);
            value = rb_str_new((const char*)bytea, bytea_len);
            PQfreemem(bytea);
            return rb_funcall(cStringIO, fnew, 1, value);
        case 20:
        case 21:
        case 23:
            return rb_cstr2inum(data, 10);
        case 18:
        case 19:
        case 25:
            return rb_enc_str_new(data, size, rb_utf8_encoding());
        case 700:
        case 701:
            return rb_float_new(atof(data));
        case 1700:
            return rb_funcall(cBigDecimal, fnew, 1, rb_str_new(data, size));
        case 1114:
        case 1184:
            return datetime_parse(cSwiftDateTime, data, size);
        case 1082:
            return date_parse(cSwiftDateTime, data, size);
        default:
            return Qnil;
    }
}

VALUE typecast_description(VALUE types) {
    VALUE strings = rb_ary_new();
    for (size_t i = 0; i < RARRAY_LEN(types); i++) {
        VALUE type = Qnil;
        int oid = NUM2INT(rb_ary_entry(types, i));
        for (size_t j = 0; j < sizeof(oid_types) / sizeof(oid_types[0]); j++) {
            if (oid_types[j].oid == oid) {
                type = rb_str_new2(oid_types[j].type);
                break;
            }
        }
        rb_ary_push(strings, type);
    }
    return strings;
}

VALUE typecast_typemap(void) {
    VALUE typemap = rb_hash_new();
    for (size_t i = 0; i < sizeof(oid_types) / sizeof(oid_types[0]); i++)
        rb_hash_aset(typemap, INT2FIX(oid_types[i].oid), rb_str_new2(oid_types[i].type));
    return typemap;
}

void init_swift_db_postgres_typecast() {
    rb_require("bigdecimal");
    rb_require("stringio");
    rb_require("date");

    cStringIO   = CONST_GET(rb_mKernel, "StringIO");
    cBigDecimal = CONST_GET(rb_mKernel, "BigDecimal");
    cDateTime   = CONST_GET(rb_mKernel, "DateTime");
    fnew        = rb_intern("new");
    fto_date    = rb_intern("to_date");
    fstrftime   = rb_intern("strftime");
    dtformat    = rb_str_new2("%F %T.%N %z");

    rb_global_variable(&dtformat);
}
